### PR TITLE
docker dev environment

### PR DESCRIPTION
**Pull request by nebulon42**
This adds a Docker dev environment which allows to run the site in a Docker container without the need to install all the Python libraries and other prerequisites such as PostgreSQL locally. On the server side it uses CherryPy which is proxied by nginX and also nginX to serve static resources.

Documentation can be found in the Wiki: https://github.com/lonvia/waymarked-trails-site/wiki/Running-Waymarked-Trails-with-Docker

Open question: Is there a way to skip the creation of the node store file or create a much smaller one? If you load a small amount of data I would expect that you do not need that or at least not 35 GB.

### CODE REVIEW

**Comment by jschleic** [comment]
I built a Docker container for development in the past and came across the same question about the node store file...

For development cherrypy can serve `/static` files so there is no need for nginx:

```diff
diff --git a/frontend.py b/frontend.py
index 065a111..35e7a4b 100755
--- a/frontend.py
+++ b/frontend.py
@@ -114,6 +114,12 @@ def setup_site(confname, script_name='', debug=False):
                                         'tools.staticfile.filename':
                                           '%s/img/map/map_%s.ico' %
                                           (globalconf['MEDIA_ROOT'], confname)
+                                        },
+                                    '/static':
+                                    {
+                                        'tools.staticdir.on': True,
+                                        'tools.staticdir.dir': '%s' % (globalconf['MEDIA_ROOT'])
```

Does this interfere with the production setup? Otherwise I'd suggest this patch for `frontend.py`

**Comment by lonvia** [comment]
The nodestore file is a hard requirement. You could hack the code to use a hash implementation instead of the flat array. That would use less space. The even easier solution is to use `osmium renumber` (from https://github.com/osmcode/osmium-tool) and renumber all node ids. Then the flatnode file will only be as large as the extract you are using.

Adding the `\static` path in the site setup is okay but it should be configurable via config/local.py and per default disabled.

File: .gitignore [hunk 1]
```diff
@@ -5,6 +5,9 @@ frontend/static/symbols/mtb/*
 frontend/static/symbols/riding/*
 frontend/static/symbols/skating/*
 frontend/static/symbols/slopes/*
+frontend/static/.webassets-cache/*
+frontend/static/packed/*
+data.osm.pbf
```
Comment (lonvia) [inline .gitignore]: The first two are okay but the third is docker specific and does not belong in this file. I would suggest to move it into the $HOME/.gitignore file of the user inside the docker instance.

File: Dockerfile [hunk 2]
```diff
@@ -0,0 +1,24 @@
+FROM ubuntu:17.04
```
Comment (lonvia) [inline Dockerfile]: Should be a problem but just for the record, the reference system is Debian Stretch.

File: config/nginx.default [hunk 3]
```diff
@@ -0,0 +1,22 @@
+# Default server configuration
```
Comment (lonvia) [inline config/nginx.default]: It would be nice if all docker-specific files could go into a docker/ subdirectory.

File: docker-startup.sh [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+do
+  pg_isready -q && echo "PostgreSQL running" && break
+  sleep 2
+  i=$((i+1))
+done
+test $i -gt $MAXCOUNT && echo "Timeout while waiting for PostgreSQL to be running"
+
+case "$1" in
+import)
+  psql -c "SELECT 1 FROM pg_database WHERE datname = 'planet';" | grep -q 1 && dropdb planet
```
Comment (lonvia) [inline docker-startup.sh]: Simply do: `dropdb --if-exists planet`

File: docker-startup.sh [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+
+  python3 makedb.py -j 4 -f data.osm.pbf db import
+  python3 makedb.py db prepare
+
+  cd  /tmp
+  curl -s http://www.nominatim.org/data/country_grid.sql.gz > country_grid.sql.gz && \
+    gunzip country_grid.sql.gz
+
+  psql -d planet -c "DROP TABLE IF EXISTS country_osm_grid" && \
+    psql -d planet -f country_grid.sql && \
```
Comment (lonvia) [inline docker-startup.sh]: If you do `zcat /home/suzuki/osm/dev/Nominatim/data/country_osm_grid.sql.gz | psql -d planet` you don't need to keep the unzipped file around.

**Comment by nebulon42** [comment]
Thanks for the comments.

> I create an organisation for waymarked-trails, move the main repo there and then create a repo for the docker installation

I'm fine with that if you are willing to do that. It will help keeping things separated but we'll have to figure out how we integrate the docker stuff and this repository here.

**Comment by lonvia** [comment]
> I'm fine with that if you are willing to do that.

Okay, will do some time next week.

> It will help keeping things separated but we'll have to figure out how we integrate the docker stuff and this repository here.

Is there anything apart from the .gitignore file that needs to reside in this repository here? It looks like it could just be checked out in the same way osgende is.

Thinking a bit further, the docker image should probably also be able to render tiles. But I see that needs some more explanation o the wiki pages.

**Comment by lonvia** [comment]
I have finally gotten around to creating the repo. You should have received an invite. Feel free to commit directly to that repo.

Closing here.

**Pull request closed without merging**